test/e2e/dra/latency.go

```diff
@@ -0,0 +1,121 @@
+	"github.com/onsi/gomega"
+	v1 "k8s.io/api/core/v1"
+	resourceapi "k8s.io/api/resource/v1alpha3"
+	"k8s.io/kubernetes/pkg/kubelet/metrics"
+	"k8s.io/kubernetes/test/e2e/feature"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2emetrics "k8s.io/kubernetes/test/e2e/framework/metrics"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	admissionapi "k8s.io/pod-security-admission/api"
+)
+
+const (
+	PrepareResourcesDurationSum   = metrics.DRAPrepareResourcesDurationKey + "_sum"
+	UnprepareResourcesDurationSum = metrics.DRAUnprepareResourcesDurationKey + "_sum"
+)
+
+var _ = framework.SIGDescribe("node")("DRA latency metrics", framework.WithSerial(), feature.DynamicResourceAllocation, func() {
+	f := framework.NewDefaultFramework("dra")
+
+	// The driver containers have to run with sufficient privileges to
+	// modify /var/lib/kubelet/plugins.
+	f.NamespacePodSecurityLevel = admissionapi.LevelPrivileged
+
+	ginkgo.Context("DRA latency", func() {
+		nodes := NewNodes(f, 1, 1)
+		driver := NewDriver(f, nodes, networkResources)
+		b := newBuilder(f, driver)
+
+		ginkgo.It("node operations should have much less latency than the total pod lifecycle", func(ctx context.Context) {
+			ginkgo.By("Getting kubelet metrics before running test pod")
+
+			ginkgo.By("Getting kubelet metrics")
+			kubeletMetrics, err := e2emetrics.GetKubeletMetrics(ctx, f.ClientSet, nodes.NodeNames[0])
+			framework.ExpectNoError(err, "Failed to get kubelet metrics")
+
+			prepareResourcesDurationBefore := kubeletMetrics[PrepareResourcesDurationSum][0].Value
+			unprepareResourcesDurationBefore := kubeletMetrics[UnprepareResourcesDurationSum][0].Value
+
+			totalDuration := runPods(ctx, f, b, 1, true)
+
+			ginkgo.By("Getting kubelet metrics after running test pod")
+			kubeletMetrics, err = e2emetrics.GetKubeletMetrics(ctx, f.ClientSet, nodes.NodeNames[0])
+			framework.ExpectNoError(err, "Failed to get kubelet metrics")
+
+			ginkgo.By("Asess total DRA call durations")
+			totalDurationPrepareResources := kubeletMetrics[PrepareResourcesDurationSum][0].Value - prepareResourcesDurationBefore
+			totalDurationUnprepareResources := kubeletMetrics[UnprepareResourcesDurationSum][0].Value - unprepareResourcesDurationBefore
+			gomega.Expect(totalDurationPrepareResources).To(gomega.BeNumerically("<=", totalDuration/15))
+			gomega.Expect(totalDurationUnprepareResources).To(gomega.BeNumerically("<=", totalDuration/40))
+
+			framework.Logf("Total durations: PrepareResources: %f sec, UnprepareResources: %f sec, pod lifecycle: %f sec", totalDurationPrepareResources, totalDurationUnprepareResources, totalDuration)
+		})
+
+		ginkgo.It("total latency overhead should be under control", func(ctx context.Context) {
+			timeWithDRA := runPods(ctx, f, b, 10, true)
+			timeWithoutDRA := runPods(ctx, f, b, 10, false)
+
+			framework.Logf("Time with DRA: %f sec, Time without DRA: %f sec", timeWithDRA, timeWithoutDRA)
+			gomega.Expect(timeWithDRA).NotTo(gomega.BeNumerically(">", timeWithoutDRA*2.5))
```

Comment:
```suggestion
			gomega.Expect(timeWithDRA).To(gomega.BeNumerically("<=", timeWithoutDRA*2.5))
```

Where does this factor come from? It seems rather arbitrary.

But that's just a nit. We shouldn't be running performance tests in a general-purpose E2E test suite. There's always the possibility that there's an outlier that suffers from random delays, in particular in a busy cluster, and then this test flakes.

If we want to test performance of DRA, then it should be using the same approach that is used for measuring other Kubernetes SLI (https://github.com/kubernetes/community/blob/master/sig-scalability/slos/pod_startup_latency.md).

FWIW, I am not sure whether we want to measure performance at this point. The KEP explicitly says that fast scheduling is not a goal because the expected use cases is around long-running pods. We can measure and optimize, but we don't have performance goals.

Reply:
> I am not sure whether we want to measure performance at this point

I removed e2e test as it caused a lot of pushback. My hope was to use it to keep DRA latency controlled. It seems that this is either too early or people are not interested or I implemented it badly.